synapse/storage/databases/main/purge_events.py

```diff
@@ -317,11 +319,36 @@ async def purge_room(self, room_id: str) -> List[int]:
         Returns:
             The list of state groups to delete.
         """
-        return await self.db_pool.runInteraction(
-            "purge_room", self._purge_room_txn, room_id
+
+        # This first runs the purge transaction with READ_COMMITTED isolation level,
+        # meaning any new rows in the tables will not trigger a serialization error.
+        # We then run the same purge a second time without this isolation level to
+        # purge any of those rows which were added during the first.
+
+        state_groups_to_delete = await self.db_pool.runInteraction(
+            "purge_room",
+            self._purge_room_txn,
+            room_id=room_id,
+            isolation_level=IsolationLevel.READ_COMMITTED,
+        )
+
+        await self.db_pool.runInteraction(
```

Comment:
Do we want to add the return value here to `state_groups_to_delete`?

Reply:
Good spot! https://github.com/matrix-org/synapse/pull/12942/commits/f895acee83496ad3cd301da6517cb7abb1774522